### PR TITLE
feat(controller): opt-in to sending pod node events as pod

### PR DESCRIPTION
**Pull request by josephmcgovern-wf**
## Summary

Because of how kubernetes aggregates events and how Argo sends both
workflow and node events with the same involved object (the workflow),
some workflow and node events can get merged into a single event, without
specifying which nodes are merged into one. This makes it challenging to
respond to phase changes for nodes that have a particular annotation
because kubernetes ignores annotations when checking if an event is
similar.

Creating an opt-in feature to send node events for nodes of type "Pod"
with the involved object set to the associated pod will force kubernetes
to send an event for each pod instead of aggregated several events into
a single event and thus make it possible build a system solely that
responds to node phase changes using Argo Workflows and Argo Events.

An alternative solution to this would be creating a `WorkflowNode`
custom resource and sending events with the workflow node itself as the
involved object, but that seemed like a bigger change than necessary.

This should provide a means to solve the problem raised in
https://github.com/argoproj/argo-workflows/issues/4136

## Testing
I tested my changes by running argo locally: `make start PROFILE=postgres` and kicking off the [hello-world](https://github.com/argoproj/argo-workflows/blob/master/examples/hello-world.yaml) workflow.
I confirmed that the expected events were sent (without sending events for the pod resource)
I then modified my code to configure the controller's `NodeEvents.SendAsPod`, and set it to true. I then re-ran `make start PROFILE=postgres` and kicked off the hello-world workflow. This time, I confirmed that node events were being sent for both the workflow and the pod.

The unit tests describe and verify this scenario well.

Checklist:

* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo-workflows/blob/master/USERS.md).

Tips:

* Your PR needs to pass the required checks before it can be approved. If the check is not required (e.g. E2E tests) it does not need to pass
* Sign-off your commits to pass the DCO check: `git commit --signoff`.
* Run `make pre-commit -B` to fix codegen or lint problems. 
* Say how how you tested your changes. If you changed the UI, attach screenshots.

### CODE REVIEW

File: workflow/controller/operator.go
Comment (kennytrytek-wf) [inline workflow/controller/operator.go:2157]: This event is still sent even if the pod event is sent, resulting in sending two similar node events with the same `message` and `reason`, but a different `involvedObject`. As your unit tests show, this results in a somewhat noisier event stream:
```
"Normal WorkflowRunning Workflow Running",
"Normal WorkflowNodeRunning Running node no-dag-or-steps",
"Normal WorkflowNodeRunning Running node no-dag-or-steps",
"Normal WorkflowNodeSucceeded Succeeded node no-dag-or-steps",
"Normal WorkflowNodeSucceeded Succeeded node no-dag-or-steps",
"Normal WorkflowSucceeded Workflow completed",
```
In the case that a pod event is sent, I think it is better if the workflow event is not sent, but the Argo maintainers may have a different opinion. I assume this was a conscious design decision; was there any particular reason you chose to send both events?

Comment (josephmcgovern-wf) [reply workflow/controller/operator.go:2157]: Yes, it was to make it backwards compatible. It felt better to accept more noise when opting in to this than to have particular events that were sending no longer sent.

**Review by alexec — COMMENTED** [review]
I get why you want this, but I wonder if there is another way to prevent events being merged into single events. Node events have annotations associated with them, which should prevent the merging. I wonder why that is not working?

What happens with a `NodeRunning` event when there is no pod to associate?

If you want to know when nodes start and finished - are there already pod events for this?

**Comment by josephmcgovern-wf** [comment]
@alexec thanks for the response.

There are pod started events, but not pod completed or pod deleted events.

Kubernetes doesn't actually consider annotations when they are determining if an event is similar enough to aggregate. [These](https://github.com/kubernetes/kubernetes/blob/c23a4b08d704156db724769a32edc71c921095d4/staging/src/k8s.io/client-go/tools/record/events_cache.go#L156-L173) are the fields that the aggregator takes into account.

If you would prefer to continue sending them to the workflow, we could modify another one of those fields, such as the reason (we could append the node name to the reason).

**Comment by alexec** [comment]
So just these three fields?
```
                event.Source.Component,
		event.ReportingController,
		event.ReportingInstance,
```
I'm not sure what these represent?

**Comment by josephmcgovern-wf** [comment]
I'm not sure I understand what you're asking. Those docs say:

> EventAggregatorByReasonFunc aggregates events by exact match on event.Source, event.InvolvedObject, event.Type, event.Reason, event.ReportingController and event.ReportingInstance

Thus, events that differ in those fields will not be aggregated. My solution sets the `involvedObject` to the pod, which would have a unique name and id. Argo currently sets the `involvedObject` to the workflow, whose name and id (and any other property within the object) is not unique across nodes.

File: config/node_events.go
Comment (alexec) [inline config/node_events.go:5]: this can be `bool` as default would be false

File: workflow/controller/operator.go
Comment (alexec) [inline workflow/controller/operator.go:2115]: Tthere is a faster way to do this. Use the `woc.controller.podInformer`. I think it is `woc.controller.podInformer.GetStore().GetKeyKey(woc.wf.Namepsace+"/"+node.ID)`

File: workflow/controller/operator.go
Comment (alexec) [reply workflow/controller/operator.go:2157]: I don't think we should send the message twice. This can already produced a high-throughput of events, and this would 2x that throughput.

For pod nodes, just send it with the pod as involved object.

File: workflow/controller/operator_test.go
Comment (alexec) [inline workflow/controller/operator_test.go:3478]: great tests!

**Review by alexec — CHANGES_REQUESTED** [review]
Design is solid. Some changes to implementation requested.

File: workflow/controller/operator.go
Comment (kennytrytek-wf) [reply workflow/controller/operator.go:2157]: > For pod nodes, just send it with the pod as involved object.

@alexec, does that mean this change would need to wait for a major Argo release? How does the backwards-incompatibility affect the release schedule?

Comment (josephmcgovern-wf) [reply workflow/controller/operator.go:2157]: I'm not sure it really would be backwards incompatible. This is an opt-in feature, so that would be a change you acknowledge by opting in.

Comment (kennytrytek-wf) [reply workflow/controller/operator.go:2157]: Yes, that's a good point. It shouldn't change anything by default.

**Comment by alexec** [comment]
Please fix test:
```
--- FAIL: TestGetPodByNode (0.10s)
    operator_test.go:3509: 
        	Error Trace:	operator_test.go:3509
        	Error:      	Expected value not to be nil.
        	Test:       	TestGetPodByNode
    operator_test.go:3510: 
        	Error Trace:	operator_test.go:3510
        	Error:      	Expected nil, but got: &errors.errorString{s:"No pod could be found for node dag-events.a"}
        	Test:       	TestGetPodByNode
```

**Comment by josephmcgovern-wf** [comment]
@alexec I've updated this branch per your requests. Let me know if you think there should be any other changes

**Review by kennytrytek-wf — APPROVED** [review]
⛄

**Review by alexec — APPROVED** [review]

**Pull request merged**
